cpp/src/groupby/hash/groupby.cu

```diff
@@ -68,8 +81,248 @@ bool constexpr is_hash_aggregation(aggregation::Kind t) {
   // return array_contains(hash_aggregations, t);
   return (t == aggregation::SUM) or (t == aggregation::MIN) or
          (t == aggregation::MAX) or (t == aggregation::COUNT) or
-         (t == aggregation::MEAN);
+         (t == aggregation::ARGMIN) or (t == aggregation::ARGMAX);
+}
+
+// flatten aggs to filter in single pass aggs 
+std::tuple<table_view, std::vector<aggregation::Kind>, std::vector<size_t> >
+flatten_single_pass_aggs(std::vector<aggregation_request> const& requests) {
+  std::vector<column_view> columns;
+  std::vector<aggregation::Kind> agg_kinds;
+  std::vector<size_t> col_ids;
+
+  for (size_t i = 0; i < requests.size(); i++) {
+    auto const& request = requests[i];
+    auto const& agg_v = request.aggregations;
+
+    auto insert_agg = [&agg_kinds, &columns, &col_ids, &request, i]
+    (aggregation::Kind k) {
+      agg_kinds.push_back(k);
+      columns.push_back(request.values);
+      col_ids.push_back(i);
+    };
+
+    std::for_each(agg_v.begin(), agg_v.end(),
```

Comment:
To me, these loops over containers that just call a lambda and have a huge capture list would be more succinct and clearer with a range for:

```suggestion
    for(auto const& agg : agg_v) {
```

Reply:
I saw the `for` and looked for an algorithm as well but it looks like `col_ids.push_back(i);` needs the explicit index. :/ This is when you ache for Python's `enumerate`

Reply:
Maybe just simplify the capture list with `[&]`.

Reply:
> Maybe just simplify the capture list with `[&]`.

I'd love to because i don't like verbosity but I've been discouraged from doing this https://github.com/rapidsai/cudf/pull/3597#discussion_r358960515

Reply:
I initially did this with a range-based for loop (and prefer it) but the only reason I changed it was because of the guideline to use stl algorithms as much as possible.

Reply:
I don't think we have an idealistic view about STL algorithms. They have their benefits in terms of declarative programming, but TBH I find a for_each that exists only to implement a loop over a container to have a lower signal to noise ratio than a range-based for loop over the same container. Lambdas are ugly, especially with @aschaffer 's capture rule...